### PR TITLE
Serialization for DataBox

### PR DESCRIPTION
**Pull request by nilsdeppe**
## Proposed changes

- serialization of boost::variant
- serialization of Deferred
- No longer sort Tags in DataBox. The reason for this is that serialization is easier if they're unsorted because then they are guaranteed to be in an order that is satisfies dependencies.
- serialization of DataBox
- fix clang-tidy 6 warnings and IWYU with boost::make_shared

### Types of changes:

- [x] Bugfix
- [x] New feature

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- Follows [code review guidelines](https://sxs-collaboration.github.io/spectre/code_review_guide.html)
- Code has documentation and unit tests
- Private member variables have a trailing underscore
- Do not use [Hungarian notation](https://en.wikipedia.org/wiki/Hungarian_notation), e.g. `double* pd_blah` is bad
- Header order:
  1. hpp corresponding to cpp (only in cpp files)
  2. Blank line (only in cpp files)
  3. STL and externals (in alphabetical order)
  4. Blank line
  5. SpECTRE includes (in alphabetical order)
- File lists in CMake are alphabetical
- Correct `noexcept` specification for functions (if unsure, mark `noexcept`)
- Mark objects `const` whenever possible
- Almost always `auto`, except with expression templates, i.e. `DataVector`
- All commits for performance changes provide quantitative evidence and the tests used to obtain said evidence.
- Make sure error messages are helpful, e.g. "The number of grid points in the matrix 'F' is not the same as the number of grid points in the determinant."
- Prefix commits addressing PR requests with `fixup`. These commits are flagged
  by Travis so we do not accidentally merge them.
- Include what you use, prefer forward declarations in hpp files.
- Explicitly make numbers floating point, e.g. `2.` or `2.0` over `2`
- Use `Tensor`'s non-member get if the indices are constant expressions, e.g.
  `get<1>(tensor)`

### CODE REVIEW

File: src/Utilities/BoostHelpers.hpp [hunk 1]
```diff
@@ -69,3 +73,18 @@ template <typename... Ts>
...
+/*!
+ * \ingroup UtilitiesGroup
+ * \brief Get the type name of the current state of the boost::variant
+ */
+template <typename... Ts>
+std::string type_of_current_state(
+    const boost::variant<Ts...>& variant) noexcept {
+  // clang-format off
+  // clang-tidy: use gsl::at (we know it'll be in bounds and want fewer
+  // includes) clang-format movoes the comment to the wrong line
```
Comment (osheamonn) [inline src/Utilities/BoostHelpers.hpp]: moves

File: src/DataStructures/DataBox/DataBox.hpp [hunk 2]
```diff
@@ -593,10 +579,18 @@ class DataBox<TagsList<Tags...>> {
    */
   using tags_list = tmpl::list<Tags...>;
 
-  using compute_item_tags =
-      tmpl::filter<tags_list, db::is_compute_item<tmpl::_1>>;
+  /// A list of all the compute items, including subitems from the compute items
+  using compute_item_tags = databox_detail::dependent_items<
+      tmpl::filter<tags_list, db::is_compute_item<tmpl::_1>>>;
+
+  /// A list of all the simple items, including subitems from the compute items
```
Comment (kidder) [inline src/DataStructures/DataBox/DataBox.hpp]: compute -> simple

File: src/DataStructures/DataBox/DataBox.hpp [hunk 3]
```diff
@@ -593,10 +579,18 @@ class DataBox<TagsList<Tags...>> {
...
+  using item_tags = tmpl::list_difference<tags_list, compute_item_tags>;
```
Comment (kidder) [inline src/DataStructures/DataBox/DataBox.hpp]: simple_item_tags

File: src/DataStructures/DataBox/DataBox.hpp [hunk 4]
```diff
@@ -593,10 +579,18 @@ class DataBox<TagsList<Tags...>> {
...
 
-  using item_tags = tmpl::remove_if<tags_list, db::is_compute_item<tmpl::_1>>;
+  /// A list of all the simple items that have subitems, including the subitems
```
Comment (kidder) [inline src/DataStructures/DataBox/DataBox.hpp]: A list of the subitems of the simple items

File: src/DataStructures/DataBox/DataBox.hpp [hunk 5]
```diff
@@ -593,10 +579,18 @@ class DataBox<TagsList<Tags...>> {
...
+  using subitem_expanded_tags = tmpl::flatten<tmpl::transform<
```
Comment (kidder) [inline src/DataStructures/DataBox/DataBox.hpp]: subitems_of_simple_items

File: src/DataStructures/DataBox/DataBox.hpp [hunk 6]
```diff
@@ -593,10 +579,18 @@ class DataBox<TagsList<Tags...>> {
...
+      tmpl::filter<item_tags,
```
Comment (kidder) [inline src/DataStructures/DataBox/DataBox.hpp]: leave out filter as Subitems is empty if it doesn't have a subitem

File: tests/Unit/Parallel/Test_AlgorithmNodelock.cpp [hunk 7]
```diff
@@ -286,8 +286,8 @@ struct NodegroupParallelComponent {
   using options = tmpl::list<>;
   using metavariables = Metavariables;
   using action_list = tmpl::list<>;
-  using initial_databox = db::DataBox<db::get_databox_list<
-      tmpl::list<Tags::total_receives_on_node, Tags::vector_of_array_indexs>>>;
+  using initial_databox = db::DataBox<
```
Comment (kidder) [inline tests/Unit/Parallel/Test_AlgorithmNodelock.cpp]: fix stuff with >=2 tags

File: src/DataStructures/DataBox/DataBox.hpp [hunk 8]
```diff
@@ -697,6 +699,19 @@ class DataBox<TagsList<Tags...>> {
   friend void mutate(DataBox<TagList>& box,                            // NOLINT
                      Invokable&& invokable, Args&&... args) noexcept;  // NOLINT
 
+  // clang-tidy: no non-const references
+  void pup(PUP::er& p) noexcept {  // NOLINT
+    using non_subitems_tags =
+        tmpl::remove_if<item_tags,
```
Comment (kidder) [inline src/DataStructures/DataBox/DataBox.hpp]: list difference?

File: tests/Unit/DataStructures/Test_DataBox.cpp [hunk 9]
```diff
@@ -1678,3 +1684,455 @@ SPECTRE_TEST_CASE("Unit.DataStructures.DataBox.Subitems",
...
+            deserialized_serialization_test_box) !=
+        &*db::get<test_subitems::Second<3>>(serialization_test_box));
+
+  counting_func<0>::count = 0;
+  counting_func<1>::count = 0;
+  test_subitems::Parent<2, true>::count = 0;
+  test_subitems::Parent<3, true, true>::count = 0;
+}
+}  // namespace
+
```
Comment (kidder) [inline tests/Unit/DataStructures/Test_DataBox.cpp:2213]: check compute item that depends upon one of the subitems

File: src/Parallel/CharmRegistration.hpp [hunk 10]
```diff
@@ -423,14 +423,18 @@ struct RegisterReductionAction : RegistrationHelper {
 template <typename Derived>
 bool register_func_with_charm() noexcept {
   if (charm_register_list_size >= charm_register_list_capacity) {
-    auto* const t =
+    // clang-tidy: use gsl::owner (we don't use raw owning pointers unless
```
Comment (kidder) [inline src/Parallel/CharmRegistration.hpp:426]: format

**Review by kidder — CHANGES_REQUESTED** [review]

File: src/DataStructures/DataBox/DataBox.hpp [hunk 11]
```diff
@@ -585,14 +585,12 @@ class DataBox<TagsList<Tags...>> {
...
+  /// A list of all the simple items, including subitems from the simple items
+  using simple_item_tags = tmpl::list_difference<tags_list, compute_item_tags>;
 
-  /// A list of all the simple items that have subitems, including the subitems
-  using subitem_expanded_tags = tmpl::flatten<tmpl::transform<
-      tmpl::filter<item_tags,
-                   tmpl::bind<databox_detail::has_subitems, tmpl::_1>>,
-      db::Subitems<tmpl::_1>>>;
+  /// A list of the subitems of the simple items
+  using subitem_expanded_tags =
```
Comment (kidder) [inline src/DataStructures/DataBox/DataBox.hpp]: could this just be subitem_tags?

Comment (nilsdeppe) [reply src/DataStructures/DataBox/DataBox.hpp]: Woops, I missed updating this. Maybe `simple_subitems_tags`?

**Review by kidder — DISMISSED** [review]

**Review by wthrowe — DISMISSED** [review]

**Comment by nilsdeppe** [comment]
I rebased this on top of develop to try and get rid of the clang-tidy and IWYU warnings in Time-related code that @wthrowe fixed in an earlier PR.

**Review by wthrowe — APPROVED** [review]

**Review by kidder — APPROVED** [review]

**Pull request merged**